### PR TITLE
Moving functions from faasm

### PR DESCRIPTION
**Pull request by Shillaker**
The vast majority of changes in this PR are formatting, addition or deletion. 

- Adding C/C++ functions that used to live in Faasm
- Removing cli.sh, docker-compose etc. that's now handled in the faasm dev environment
- Matched formatting with Faasm repo

The only changes that aren't copy-paste, removal or formatting are:

- [.github/workflows/test.yml](https://github.com/faasm/cpp/pull/31/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957f) - new build step for wasm
- [faasmtools/env.py](https://github.com/faasm/cpp/pull/31/files#diff-b9e6e0cc0a67e5174486b4075c5e8386b9346041b99290ba3afac3c259d6b1ad) - added a couple of constants
- [tasks/func.py](https://github.com/faasm/cpp/pull/31/files#diff-825aa52b0be441a94a1c29652bab306c854dc32cefa775d2dd0728c51ec425ce) - changing directories
- [tasks/libfake.py](https://github.com/faasm/cpp/pull/31/files#diff-217e28989d3b6b8618e40a2067d6e9db7accac9f4d0c7454ebefdd8e8d149868) - changing directories

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+---
+Language: Cpp
+
+# Avoid excessive linebreaks on function definitions
+AlwaysBreakAfterDefinitionReturnType: None
+AlwaysBreakAfterReturnType: None
+
+IndentWidth: 4
+DerivePointerAlignment: false
+---
```
Comment (Shillaker) [inline .clang-format:13]: For some reason the formatting rules were different in this repo so all the new code was failing. Adding this fixes it but changes the formatting in other files. I'll comment where relevant.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -65,6 +65,26 @@ jobs:
       - name: "Run the tests"
         run: /build/cpp/static/bin/tests
   
+  wasm:
```
Comment (Shillaker) [inline .github/workflows/tests.yml:68]: This just checks that the wasm compiles. We'll still check its execution in the faasm tests

File: emulator/emulator.cpp [hunk 3]
```diff
@@ -1,5 +1,6 @@
 
-extern "C" {
```
Comment (Shillaker) [inline emulator/emulator.cpp:2]: This file is just formatting.

File: emulator/faasm/emulator.h [hunk 4]
```diff
@@ -6,15 +6,15 @@
 
 #include <faabric/state/State.h>
```
Comment (Shillaker) [inline emulator/faasm/emulator.h:7]: Formatting

File: examples/native_exe.cpp [hunk 5]
```diff
@@ -1,11 +1,12 @@
 #include "native_lib.h"
```
Comment (Shillaker) [inline examples/native_exe.cpp:1]: Formatting

File: faasmtools/compile_util.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+from subprocess import run
```
Comment (Shillaker) [inline faasmtools/compile_util.py:1]: This was moved from the Faasm repo, where it's no longer relevant

File: faasmtools/env.py [hunk 7]
```diff
@@ -3,6 +3,8 @@
 
 PROJ_ROOT = dirname(dirname(abspath(__file__)))
 THIRD_PARTY_DIR = join(PROJ_ROOT, "third-party")
+WASM_DIR = "/usr/local/faasm/wasm"
+FAASM_RUNTIME_ROOT = "/usr/local/faasm/runtime_root"
```
Comment (Shillaker) [inline faasmtools/env.py:7]: This dir is now hard-coded in quite a few places, but it's preferable to having Python dependencies across all the modules just for these constants.

File: func/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,92 @@
+cmake_minimum_required(VERSION 3.8.0)
```
Comment (Shillaker) [inline func/CMakeLists.txt:1]: Copied in from Faasm

File: func/demo/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,122 @@
+set(FAASM_USER demo)
```
Comment (Shillaker) [inline func/demo/CMakeLists.txt:1]: Copied from Faasm

File: faasmtools/env.py (hunk 7 above)
Comment (csegarragonz) [reply faasmtools/env.py:7]: Indeed, this was making `inv -l` fail in so many places unnecessarily.

**Review by csegarragonz — APPROVED** [review]

**Pull request merged**